topology/probes/k8s/cluster.go

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2018 IBM Corp.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package k8s
+
+import (
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/probe"
+	"github.com/skydive-project/skydive/topology"
+	"github.com/skydive-project/skydive/topology/graph"
+
+	"k8s.io/client-go/kubernetes"
+)
+
+// ClusterName is the name of the k8s cluster
+const ClusterName = "cluster"
+
+type clusterCache struct {
+	*graph.EventHandler
+	graph *graph.Graph
+}
+
+func (c *clusterCache) Start() {
+	c.graph.Lock()
+	defer c.graph.Unlock()
+
+	m := NewMetadata(Manager, "cluster", nil, ClusterName)
+	node := c.graph.NewNode(graph.GenID(), m, "")
+	c.NotifyEvent(graph.NodeAdded, node)
+	logging.GetLogger().Debugf("Added cluster{Name: %s}", ClusterName)
+}
+
+func (c *clusterCache) Stop() {
+}
+
+func newClusterProbe(clientset *kubernetes.Clientset, g *graph.Graph) Subprobe {
+	return &clusterCache{
+		EventHandler: graph.NewEventHandler(100),
+		graph:        g,
+	}
+}
+
+func newClusterLinker(g *graph.Graph, subprobes map[string]Subprobe, manager, ty string) probe.Probe {
+	dstCache := subprobes[ty]
+	if dstCache == nil {
+		return nil
+	}
+
+	clusterIndexer := graph.NewMetadataIndexer(g, g, graph.Metadata{"Manager": Manager, "Type": "cluster"})
```

Comment:
Creating a indexer for a single resource is probably overkill

Reply:
it simplified implementation (over sharing the indexer between all the cluster linkers) and I don't think there is a performance penalty

Reply:
Yes, there is. For every change on the graph, one of OnNodeAdded/OnNodeUpdated/OnEdgeAdded/OnEdgeUpdated will be called. The only time this will do something is when we add the cluster node, and we exactly know when this will happen. I'm sorry but it makes no sense to index a single resource

Reply:
The previous implementation did not make use of an  MetadataIndexerLinker for the links between cluster and the top level objects. A `classic` linker that just implements GetBALinks is enough.

Reply:
@lebauce - for have changed this a little such that a use a dedicated clusterEventHandler (so as not to be tirggered on every add/del/update of a node on the graph).